hnn_core/parallel_backends.py

```diff
@@ -351,8 +393,23 @@ def simulate(self, net):
         self.sel.register(pipe_stderr_r, selectors.EVENT_READ,
                           self._read_stdout)
 
+        data_len = 0
+        completed = False
+        timeout = 0
         # loop while the process is running
-        while proc.poll() is None:
+        while True:
+            if not proc.poll() is None:
+                if completed is True:
+                    break
+                elif timeout > 4:
+                    # This is indicative of a failure. For debugging purposes.
+                    warn("Timed out (5s) waiting for end of data after child "
+                         "process stopped")
+                    break
+                else:
+                    timeout += 1
+                    sleep(1)
```

Comment:
isn't this part of some system library? I haven't looked deeply but it feels like [Popen.wait](https://docs.python.org/3/library/subprocess.html#subprocess.Popen.wait) might do the same?

Reply:
The problem with calls like `Popen.wait()` is that they block. In this case, the checks are for loop control... and the combination of conditions wouldn't match a typical system library. I'm making two specific checks if the loop should be continued and neither should block.

Reply:
> very dumb question, what's the reason for encoding the data as base64? I just googled and found this. One of the answers says: "Base64 is just an inefficient method to store binary data." I trust you on this, I am just curious what is the reason you pickle and then encode the data as base 64? What are the pros/cons of different choices (if any)?

No, good question. Starting with the limitations: stdout and stderr can only pass strings. Pikle returns a string, except it's in the 'latin-1' encoding. Base64 is a way to convert the byte string into an ASCII encoding (i.e. portable). I remember finding support for using this method on stackoverflow when sending pickled data over a network and particularly the `cloudpickle` project (same `Popen` use case as us). However, I rechecked that project and they are no longer using baseX encoding! It looks like the reason for base32/base64 was for python2/windows compatibility. I'm interpreting python3 with windows as being okay.

https://github.com/cloudpipe/cloudpickle/commit/65b240866c7d861e79fcc68dc68ce06f3e1d6f57#diff-9d24dc3dc9330ef153577f547b4024b50a095ce91290796e44b7a59dcd7b9119L49-R105

Straight up pickle loads and dumps should work!

Reply:
> The problem with calls like Popen.wait() is that they block.

So just to complete my understanding, you don't want them to block because of the GUI?

Reply:
> > The problem with calls like Popen.wait() is that they block.
> 
> So just to complete my understanding, you don't want them to block because of the GUI?

Even, with the command-line. If a `wait()` call is made, then simulation progress isn't printed to the screen until `wait()` returns. Same reason as for the GUI.

Reply:
> Straight up pickle loads and dumps should work!

Unfortunately, not the case. I tried with several different ways, but could not successfully unpickle the data without using base64 encoding. Here's a similar issue:
https://stackoverflow.com/questions/30469575/how-to-pickle-and-unpickle-to-portable-string-in-python-3

I will keep using base64 for this PR